tools/tplgtool.py

```diff
@@ -599,17 +600,136 @@ def _merge_pcm_list(self, pcm_block_list):
                 merged_pcm_list.append(pcm)
         return merged_pcm_list
 
+    # graph node form: {"name":name, "widget":widget, "ctrl":ctrl, "source":source, "sink":sink}
+    # return values:
+    #   link_head_list: head node list of every graph
+    #   node_list: list of all nodes
+    def _link_graph(self):
+        node_list = self._init_node_list()
+        if node_list == []:
+            print("No widget in topology!")
+            sys.exit(1)
+
+        for graphs in self._tplg["graph_list"]:
+            # const variables for graph
+            SOURCE = 0
+            CONTROL = 1
+            SINK = 2
+
+            for graph in graphs:
+                source_node = self.find_node_by_name(graph[SOURCE], node_list)
+                sink_node = self.find_node_by_name(graph[SINK], node_list)
+
+                # some of the names in graph are from widget["name"] (eg: PCM0P), and others are
+                # from widget["sname"] (eg: SSP1.OUT), use the name in graph as standard name of a node
+                if source_node["name"] != graph[SOURCE]:
+                    source_node["name"] = graph[SOURCE]
+                if sink_node["name"] != graph[SINK]:
+                    sink_node["name"] = graph[SINK]
+
+                source_node["ctrl"] = graph[CONTROL]
+                # a node may link to multiple nodes
+                if source_node["sink"] != None: # this node has a link already
+                    # if already link to a node
+                    if type(source_node["sink"]) != list:
+                        next_list = [source_node["sink"], sink_node]
+                        source_node["sink"] = next_list
+                    # if already link to a node list
+                    else:
+                        source_node["sink"].append(sink_node)
+                # not link to any node
+                else:
+                    source_node["sink"] = sink_node
```

Comment:
change source_node["slink"] as list at init  
here you just need to do this:  
`source_node["sink"].append(sink_node)`  
because `sink_node` couldn't empty